hw/ip/otbn/util/shared/control_flow.py

```diff
@@ -0,0 +1,392 @@
+        pcs_to_print = range(sec.start, sec.end + 4, 4)
+        if len(symbols) > 0:
+            out.append((0, '<{}>'.format(', '.join(symbols))))
+            if entry_pc in already_printed and not concise:
+                out.append((0, '(see above)'))
+                pcs_to_print = []
+        if concise:
+            # only print last insn
+            out.append((0, '...'))
+            pcs_to_print = [sec.end]
+        for pc in pcs_to_print:
+            insn = program.get_insn(pc)
+            operands = program.get_operands(pc)
+            out.append((0, '0x{:x}: {}'.format(pc,
+                                               insn.disassemble(pc,
+                                                                operands))))
+            already_printed.add(sec.start)
+        # Indent if we have more than 1 non-special edge
+        non_special_edges = [
+            loc for loc in self.edges[entry_pc] if not loc.is_special()
+        ]
+        child_indent = 0 if len(non_special_edges) <= 1 else 2
+        for loc in self.edges[entry_pc]:
+            out.append((0, '->'))
+            if loc.is_special():
+                out.append((child_indent, loc.pretty()))
+                if isinstance(loc, Ret) and len(call_stack) > 0:
+                    # pop the call stack and print flow after return to caller
+                    out.extend([(indent - 2, line)
+                                for indent, line in self._pretty_lines(
+                                    program, call_stack[0], concise,
+                                    already_printed, call_stack[1:])])
+                if isinstance(loc, LoopStart):
+                    # Prevent loops in printing by checking we didn't already print this loop
+                    if loc.loop_start_pc not in already_printed:
+                        out.extend([(indent + child_indent, line)
+                                    for indent, line in self._pretty_lines(
+                                        program, loc.loop_start_pc, concise,
+                                        already_printed, call_stack)])
+            else:
+                insn = program.get_insn(sec.end)
+                operands = program.get_operands(sec.end)
+                if insn.mnemonic == 'jal' and operands['grd'] == 1:
+                    # push to to the call stack
+                    call_stack = [sec.end + 4] + call_stack
+                    child_indent += 2
+                out.extend([
+                    (indent + child_indent, line)
+                    for indent, line in self._pretty_lines(
+                        program, loc.pc, concise, already_printed, call_stack)
+                ])
+        return out
+
+    def pretty(self, program, indent=0, concise=False):
+        '''Returns string for pretty-printing.'''
+        lines = self._pretty_lines(program, self.start, concise, set())
+        lines = [(max(0, indent + lindent), line) for (lindent, line) in lines]
```

Comment:
Is this max ever needed? And if it fires, won't it make a bit of a mess of the indentation?